### ISSUE TITLE
CMake Build Issues on OSx

### ISSUE THREAD

**Issue by ambasta** [google/or-tools#406]
This is a bug to track build issues on OSx while using cmake.

The currently known issue is while configuring the project as:

-- The CXX compiler identification is AppleClang 8.1.0.8020042
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/local/bin/pkg-config (found version "0.29.2")
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Checking for module 'libglog>=0.3.5'
-- Found libglog, version 0.3.5
-- Checking for module 'cbc'
-- Found cbc, version 2.9.6
CMake Error at CMakeLists.txt:79 (ADD_DEPENDENCIES):
add_dependencies called with incorrect number of arguments

PROJECT_NAME seems undefined

**Comment by lperron** [comment]
running cmake went fine. 
It seems something is wrong with CBC config file (it read the brew version).

/Users/lperron/Work/cmake-or-tools/ortools/linear_solver/cbc_interface.cc:35:10: fatal error: 
      'coin/CbcConfig.h' file not found
#include "coin/CbcConfig.h"

**Comment by ambasta** [comment]
Should I assume you ran "brew install cbc" to get cbc via brew?

**Comment by lperron** [comment]
yes.

Looking at CMakeCache.txt, it seems that or-tools does not add the cbc paths to cflags.

[CMakeCache.txt](https://github.com/google/or-tools/files/1054532/CMakeCache.txt)

**Comment by lperron** [comment]
Actually, looking at build/ortools/linear_solver/.../flags.cmake, the -I<path to CBC> is there.
So I do not know.

**Comment by ambasta** [comment]
I've been trying to install cbc via brew, but it keeps throwing configure: error: Required package CoinUtils not available. I've however udpated the build a little which might hopefully address your issues

**Comment by ambasta** [comment]
Hey @lperron 

So I figured out the first issue with system-wide cbc on OSX. pkg-config on brew installed cbc throws the following

```
pkg-config --cflags-only-I cbc
-I/usr/local/Cellar/cbc/2.9.6/include/cbc/coin -I/usr/local/Cellar/cgl/0.59.6/include/cgl/coin -I/usr/local/Cellar/clp/1.16.7/include/clp/coin -I/usr/local/Cellar/osi/0.107.5/include/osi/coin -I/usr/local/Cellar/coinutils/2.10.10/include/coinutils/coin

```
At this point, while `#include <CbcConfig.h>` will succeed `#include <coin/CbcConfig.h>` fails since coin subdir is already part of the include path.

How are you specifying include paths in the current make tool?

**Comment by lperron** [comment]
good catch.

As I wrote on the last merge request, I cleaned everything from brew.
It does not build gflags, and forget to link with -lz.

Still, I believe we (you) are making progress.

Thanks

**Comment by lperron** [comment]
on the include part, is there a flag that tells use we are building using cmake?

**Comment by lperron** [comment]
I changed the gflags part in CMakeLists.txt to mirror the glog part:

    SET(GFLAGS_VERSION ${Gflags_VERSION})
    PKG_CHECK_MODULES(GFLAGS libgflags>=${GFLAGS_VERSION})
    # Set again since pkg_check_module resets GFLAGS_VERSION
    SET(GFLAGS_VERSION ${Gflags_VERSION})
    IF(NOT GFLAGS_FOUND)
        MESSAGE(STATUS "Did not find system gflags. Building as an external project.")
        INCLUDE(cmake/external/gflags.cmake)
    ENDIF()
    INCLUDE_DIRECTORIES(${GFLAGS_INCLUDE_DIRS})


Now it builds gflags.

**Comment by ambasta** [comment]
Okay, found the issue with gflags as well. gflags ships its CMakeConfig file with project name gflags, not Gflags.. That is case sensitive..

**Comment by lperron** [comment]
ok.

BTW, glog should build using gflags. so it should not start build glog before gflags is installed.

**Comment by ambasta** [comment]
Will specify gflags as a dependency for glog then :)

**Comment by lperron** [comment]
Hi, 

I pushed a working fix on my mac.
Hopefully, it did not break anything.

The next step would be to finalise the C++ build on unix. I believe we just need to put version numbers on libraries. 

There are two parts to it:
  - Major and Minor are stored in <or-tools>/Makefile, but we could create a Version.txt file for it.
  - revision number is actually the size of the github revision list (see GIT_REVISION in makefiles/Makefile.port).

Then we can move forward on 3 fronts:
  - python
  - windows build, in particular creating a solution to build the whole project.
  - packaging (mac package, linux packages, python modules...).

Later, java, C# if possible...

Thank you so much for your help.

**Comment by lperron** [comment]
Another thing to do, add the include files of glog, gflags, and protobuf in the archive.

BTW, I was thinking about the discovery process of dependencies.
I would be in favor of building glog, gflags, protobuf and cbc by default, unless the user add some special flags that will tell cmake to use already installed packages, and in that case they will be required.

**Comment by ambasta** [comment]
Sure, we can go the firefox route of system-cbc, system-glog etc.

Second, I agree with the versioning structure for the project. However, I'd recommend either picking it up from the git tag against current commit (for releases) or fallback to a VERSION file (which on master should always point to a dev release?)

Finally, if we do include the headers of the dependencies, then we'd have to install them in subdir of /usr/include/ortools/ to avoid conflicts with system wide installs for the same.

**Comment by lperron** [comment]
version: 
I like MAJOR.MINOR.REVISION as a versioning number.

I need MAJOR and MINOR to be hand controlled to mark new releases, and indicates if I break compatibility (bump in MAJOR).

About the includes, 
I would like 

include/
  coin/
  gflags/
  ortools/

And I can document that they should be include with #include " " such that they are found before system paths.

PS: Can you send my your email (by mail lperron@google.com). It will be useful.

**Comment by mjfwest** [comment]
Trying the cmake branch on my mac, OS X 10.12.5 (16F73),  I get the following error:
```
CMake Error at /usr/local/Cellar/cmake/3.8.2/share/cmake/Modules/ExternalProject.cmake:2005 (message):
   No download info given for 'bison-3.0.4' and its source directory:

    /Users/xxxxx/Documents/repositories/or-tools/build/../dependencies/Source/bison-3.0.4

   is not an existing non-empty directory.  Please specify one of:

    * SOURCE_DIR with an existing non-empty directory
    * URL
    * GIT_REPOSITORY
    * HG_REPOSITORY
    * CVS_REPOSITORY and CVS_MODULE
    * SVN_REVISION
    * DOWNLOAD_COMMAND
 Call Stack (most recent call first):
   /usr/local/Cellar/cmake/3.8.2/share/cmake/Modules/ExternalProject.cmake:2560 (_ep_add_download_command)
   src/flatzinc/dependencies.cmake:11 (ExternalProject_Add)
   src/flatzinc/CMakeLists.txt:1 (include)



 CMake Error at src/flatzinc/CMakeLists.txt:5 (message):
   flex could not be found.  The minimum required version is 2.5.35.
```

Is this a known bug? Do you want another issue created for this one?

**Comment by lperron** [comment]
We do not use the cmake branch. The code is head.


Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-06-14 13:18 GMT+02:00 Martin West <notifications@github.com>:

**Comment by mjfwest** [comment]
Ok. Thanks for the clarification, I'll try head.

**Comment by mjfwest** [comment]
Using cmake on master I get as far as:
```
[ 72%] Building CXX object ortools/linear_solver/CMakeFiles/ortools_linear_solver.dir/cbc_interface.cc.o
/Users/xxxxx/Documents/repositories/or-tools/ortools/linear_solver/cbc_interface.cc:35:10: fatal error: 'CbcConfig.h' file not found
#include "CbcConfig.h"
         ^
1 error generated.
make[2]: *** [ortools/linear_solver/CMakeFiles/ortools_linear_solver.dir/cbc_interface.cc.o] Error 1
make[1]: *** [ortools/linear_solver/CMakeFiles/ortools_linear_solver.dir/all] Error 2
make: *** [all] Error 2
```

Is this a known conflict with brew installed CBC? Is there a way for them to both exist?

**Comment by ambasta** [comment]
Martin

Could you please clone my fork and see if that works for you?

**Comment by mjfwest** [comment]
Hello @ambasta,
It seems that change works. It now gets further to:
```
Scanning dependencies of target ortools
[ 77%] Linking CXX shared library libortools.dylib
```
then follows thousands of lines of
`duplicate symbol __SomeSymbol in: SomeFile`
followed by:
```
ld: 2692 duplicate symbols for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [libortools.dylib] Error 1
make[1]: *** [CMakeFiles/ortools.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by ambasta** [comment]
Hey @mjfwest Would it be possible to attach a full build log. I'd recommend opening an issue against the fork instead though.

**Comment by ambasta** [comment]
I'm closing this bug unless new OSX issues are highlighted :)
